File: experiments/issue-314/ports/probe.php
<?php
// Times the PHP parser on the nesting shapes from #314.
// Usage: php probe.php (from the repository root)
spl_autoload_register(function (string $class): void {
    $prefix = 'LinkFoundation\\LinksNotation\\';
    if (str_starts_with($class, $prefix)) {
        require __DIR__ . '/../../../php/src/' . str_replace('\\', '/', substr($class, strlen($prefix))) . '.php';
    }
});
use LinkFoundation\LinksNotation\Parser;

$shapes = [
    'closed' => fn (int $d) => str_repeat('(', $d) . 'a' . str_repeat(')', $d),
    'value after' => fn (int $d) => str_repeat('(', $d) . 'a' . str_repeat(') b', $d),
    'unclosed' => fn (int $d) => str_repeat('(', $d) . 'a',
    'indented' => fn (int $d) => implode("\n", array_map(fn ($i) => str_repeat(' ', $i) . '(a', range(0, $d - 1))),
];
$parser = new Parser();
foreach ($shapes as $name => $shape) {
    printf('%-12s', $name);
    foreach ([2, 4, 8, 12, 16, 20, 24, 64, 256] as $d) {
        $start = microtime(true);
        try {
            $parser->parse($shape($d));
            $outcome = 'ok';
        } catch (Throwable $e) {
            $outcome = get_class($e);
        }
        $elapsed = microtime(true) - $start;
        printf(' %d: %.1f ms (%s),', $d, $elapsed * 1000, $outcome);
        if ($elapsed > 3) {
            break;
        }
    }
    echo "\n";
}

File: experiments/issue-314/ports/probe.py
"""Times the Python parser on the nesting shapes from #314.

Usage: python3 probe.py (run from the repository root or with PYTHONPATH=python)
"""
import sys
import time

sys.path.insert(0, "python")
from links_notation import Parser  # noqa: E402

SHAPES = {
    "closed": lambda d: "(" * d + "a" + ")" * d,
    "value after": lambda d: "(" * d + "a" + ") b" * d,
    "unclosed": lambda d: "(" * d + "a",
    "indented": lambda d: "\n".join(" " * i + "(a" for i in range(d)),
}
BUDGET = 3.0
parser = Parser()
for name, shape in SHAPES.items():
    results = []
    for depth in (2, 4, 8, 12, 16, 20, 24, 64, 256):
        start = time.perf_counter()
        try:
            parser.parse(shape(depth))
            outcome = "ok"
        except Exception as error:  # noqa: BLE001
            outcome = type(error).__name__
        elapsed = time.perf_counter() - start
        results.append(f"{depth}: {elapsed * 1000:.1f} ms ({outcome})")
        if elapsed > BUDGET:
            break
    print(f"{name:12} " + ", ".join(results))
